File: python/confuse-common/src/confuse_common/events/episode.py
"""
Graphify Episode — Canonical Data Contract

Defines the standardised episode schema that all ConFuse microservices use
when emitting processed data to Graphify. This is the canonical handoff unit
between ingestion, embedding, and the Graphify knowledge graph.

Episode = chunk + metadata + embedding (optional at emission time).
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── Source type taxonomy ──────────────────────────────────────────────────────

class EpisodeSourceType(str, Enum):
    """Content source categories for Graphify episodes."""
    CODEBASE = "codebase"
    DOCUMENT = "document"
    WEB = "web"
    VIDEO = "video"
    AUDIO = "audio"
    IMAGE = "image"
    CHAT = "chat"
    MESSAGE = "message"
    UNKNOWN = "unknown"


class EpisodeChunkType(str, Enum):
    """Granularity / semantic type of the chunk."""
    # Codebase types
    FUNCTION = "function"
    CLASS = "class"
    MODULE = "module"
    IMPORT = "import"
    CODE_BLOCK = "code_block"
    DOCSTRING = "docstring"
    # Document types
    SECTION = "section"
    PARAGRAPH = "paragraph"
    SENTENCE = "sentence"
    TABLE = "table"
    CITATION = "citation"
    # Web types
    PAGE_OVERVIEW = "page_overview"
    WEB_SECTION = "web_section"
    WEB_METADATA = "web_metadata"
    # Media types
    VIDEO_SEGMENT = "video_segment"
    AUDIO_SEGMENT = "audio_segment"
    KEYFRAME = "keyframe"
    TRANSCRIPT = "transcript"
    # Chat types
    THREAD = "thread"
    MESSAGE_TURN = "message_turn"
    # Generic
    MIXED = "mixed"
    RAW = "raw"


# ── Sub-models ────────────────────────────────────────────────────────────────

class EpisodeMetadata(BaseModel):
    """
    Content-type-specific metadata attached to every episode.

    Fields are optional because not all content types produce all metadata.
    """
    mime_type: Optional[str] = None
    filename: Optional[str] = None
    author: Optional[str] = None
    language: Optional[str] = None
    # Document-specific
    title: Optional[str] = None
    page_number: Optional[int] = None
    heading: Optional[str] = None
    # Code-specific
    symbol_name: Optional[str] = None
    line_start: Optional[int] = None
    line_end: Optional[int] = None
    complexity: Optional[int] = None
    # Media-specific
    duration_ms: Optional[int] = None
    width: Optional[int] = None
    height: Optional[int] = None
    exif: Optional[Dict[str, Any]] = None
    transcript: Optional[str] = None
    # Chat-specific
    participant: Optional[str] = None
    thread_id: Optional[str] = None
    # Entity hints for downstream graph construction
    entity_hints: List[Dict[str, Any]] = Field(default_factory=list)
    # Relationship hints for downstream graph construction
    relationship_hints: List[Dict[str, Any]] = Field(default_factory=list)
    # Arbitrary key-value pairs for extensibility
    extra: Dict[str, Any] = Field(default_factory=dict)


class EpisodeProvenance(BaseModel):
    """
    Provenance tracking — which service produced this episode, when, and how.
    Enables auditability and replay from any point in the pipeline.
    """
    origin_service: str = Field(
        ...,
        description="Service that created this episode (e.g. 'data-connector', 'unified-processor')",
    )
    version: str = Field(
        default="1.0.0",
        description="Schema version for forward/backward compatibility",
    )
    pipeline_run_id: Optional[str] = Field(
        default=None,
        description="ID of the pipeline run that created this episode",
    )
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        description="Timestamp when this episode was created",
    )
    parent_episode_id: Optional[str] = Field(
        default=None,
        description="ID of the parent episode if this is a derived/split episode",
    )


# ── Canonical Episode ─────────────────────────────────────────────────────────

class GraphifyEpisode(BaseModel):
    """
    Canonical episode contract for Graphify ingestion.

    This is the standardised handoff unit used across all ConFuse services:
    - data-connector → Kafka → Graphify
    - embeddings-service → Kafka → Graphify
    - unified-processor → Graphify (direct or Kafka)

    Fields are designed to be content-type-agnostic while supporting
    rich metadata for every supported source type.
    """

    # ── Identity ──────────────────────────────────────────────────────────
    id: str = Field(
        default_factory=lambda: str(uuid.uuid4()),
        description="Globally unique episode identifier",
    )
    source_type: EpisodeSourceType = Field(
        ...,
        description="Content source category",
    )
    source_id: str = Field(
        ...,
        description="Identifier of the source (repo ID, document ID, URL, etc.)",
    )

    # ── Content ───────────────────────────────────────────────────────────
    chunk_text: str = Field(
        ...,
        description="The actual text content of this episode chunk",
    )
    chunk_type: EpisodeChunkType = Field(
        default=EpisodeChunkType.RAW,
        description="Semantic type / granularity of this chunk",
    )

    # ── Embedding (populated by embeddings-service) ───────────────────────
    embedding: Optional[List[float]] = Field(
        default=None,
        description="Float vector embedding — populated after embedding generation",
    )
    embedding_model: Optional[str] = Field(
        default=None,
        description="Name of the model used to generate the embedding",
    )
    embedding_dimension: Optional[int] = Field(
        default=None,
        description="Dimensionality of the embedding vector",
    )

    # ── Context ───────────────────────────────────────────────────────────
    language: Optional[str] = Field(
        default=None,
        description="Programming or natural language of the content",
    )
    content_path: Optional[str] = Field(
        default=None,
        description="File path, URL, or location within the source",
    )
    group_id: Optional[str] = Field(
        default=None,
        description="Logical grouping / tenant ID for graph partitioning",
    )

    # ── Timestamps ────────────────────────────────────────────────────────
    start_timestamp: Optional[datetime] = Field(
        default=None,
        description="Start time for temporal content (video/audio segments)",
    )
    end_timestamp: Optional[datetime] = Field(
        default=None,
        description="End time for temporal content (video/audio segments)",
    )

    # ── Rich metadata ─────────────────────────────────────────────────────
    metadata: EpisodeMetadata = Field(
        default_factory=EpisodeMetadata,
        description="Content-type-specific metadata",
    )

    # ── Provenance ────────────────────────────────────────────────────────
    provenance: EpisodeProvenance = Field(
        ...,
        description="Origin and version tracking",
    )

    # ── Security / Governance ─────────────────────────────────────────────
    pii_detected: bool = Field(
        default=False,
        description="Whether PII was detected in this episode's content",
    )
    acl_tags: List[str] = Field(
        default_factory=list,
        description="Access control tags for graph query authorization",
    )

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat(),
        }

    def to_kafka_payload(self) -> dict:
        """Serialize to a Kafka-friendly dict (JSON-serializable)."""
        return self.model_dump(mode="json")

    @classmethod
    def from_kafka_payload(cls, payload: dict) -> "GraphifyEpisode":
        """Deserialize from a Kafka payload dict."""
        return cls.model_validate(payload)

File: rust/confuse-common/src/events/episode.rs
//! Graphify Episode — Canonical Data Contract (Rust)
//!
//! Defines the standardised episode schema that all ConFuse Rust services use
//! when emitting processed data to Graphify. This mirrors the Python model in
//! `confuse_common.events.episode` for cross-language consistency.
//!
//! Episode = chunk + metadata + embedding (optional at emission time).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ── Source type taxonomy ─────────────────────────────────────────────────────

/// Content source categories for Graphify episodes.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum EpisodeSourceType {
    Codebase,
    Document,
    Web,
    Video,
    Audio,
    Image,
    Chat,
    Message,
    Unknown,
}

impl std::fmt::Display for EpisodeSourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Codebase => write!(f, "codebase"),
            Self::Document => write!(f, "document"),
            Self::Web => write!(f, "web"),
            Self::Video => write!(f, "video"),
            Self::Audio => write!(f, "audio"),
            Self::Image => write!(f, "image"),
            Self::Chat => write!(f, "chat"),
            Self::Message => write!(f, "message"),
            Self::Unknown => write!(f, "unknown"),
        }
    }
}

/// Granularity / semantic type of the chunk.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EpisodeChunkType {
    // Codebase types
    Function,
    Class,
    Module,
    Import,
    CodeBlock,
    Docstring,
    // Document types
    Section,
    Paragraph,
    Sentence,
    Table,
    Citation,
    // Web types
    PageOverview,
    WebSection,
    WebMetadata,
    // Media types
    VideoSegment,
    AudioSegment,
    Keyframe,
    Transcript,
    // Chat types
    Thread,
    MessageTurn,
    // Generic
    Mixed,
    Raw,
}

impl Default for EpisodeChunkType {
    fn default() -> Self {
        Self::Raw
    }
}

// ── Sub-models ───────────────────────────────────────────────────────────────

/// Content-type-specific metadata attached to every episode.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EpisodeMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub author: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    // Document-specific
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_number: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub heading: Option<String>,
    // Code-specific
    #[serde(skip_serializing_if = "Option::is_none")]
    pub symbol_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line_start: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line_end: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub complexity: Option<u32>,
    // Media-specific
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exif: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transcript: Option<String>,
    // Chat-specific
    #[serde(skip_serializing_if = "Option::is_none")]
    pub participant: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thread_id: Option<String>,
    // Entity hints for downstream graph construction
    #[serde(default)]
    pub entity_hints: Vec<serde_json::Value>,
    // Relationship hints for downstream graph construction
    #[serde(default)]
    pub relationship_hints: Vec<serde_json::Value>,
    // Arbitrary key-value pairs for extensibility
    #[serde(default)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

/// Provenance tracking — which service produced this episode, when, and how.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EpisodeProvenance {
    /// Service that created this episode (e.g. "data-connector", "unified-processor").
    pub origin_service: String,
    /// Schema version for forward/backward compatibility.
    #[serde(default = "default_version")]
    pub version: String,
    /// ID of the pipeline run that created this episode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pipeline_run_id: Option<String>,
    /// Timestamp when this episode was created.
    #[serde(default = "Utc::now")]
    pub created_at: DateTime<Utc>,
    /// ID of the parent episode if this is a derived/split episode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_episode_id: Option<String>,
}

fn default_version() -> String {
    "1.0.0".to_string()
}

impl Default for EpisodeProvenance {
    fn default() -> Self {
        Self {
            origin_service: "unknown".to_string(),
            version: default_version(),
            pipeline_run_id: None,
            created_at: Utc::now(),
            parent_episode_id: None,
        }
    }
}

// ── Canonical Episode ────────────────────────────────────────────────────────

/// Canonical episode contract for Graphify ingestion.
///
/// This is the standardised handoff unit used across all ConFuse services:
/// - data-connector → Kafka → Graphify
/// - embeddings-service → Kafka → Graphify
/// - unified-processor → Graphify (direct or Kafka)
///
/// Fields are designed to be content-type-agnostic while supporting
/// rich metadata for every supported source type.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphifyEpisode {
    // ── Identity ─────────────────────────────────────────────────────────
    /// Globally unique episode identifier.
    pub id: String,
    /// Content source category.
    pub source_type: EpisodeSourceType,
    /// Identifier of the source (repo ID, document ID, URL, etc.).
    pub source_id: String,

    // ── Content ──────────────────────────────────────────────────────────
    /// The actual text content of this episode chunk.
    pub chunk_text: String,
    /// Semantic type / granularity of this chunk.
    #[serde(default)]
    pub chunk_type: EpisodeChunkType,

    // ── Embedding (populated by embeddings-service) ──────────────────────
    /// Float vector embedding — populated after embedding generation.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding: Option<Vec<f32>>,
    /// Name of the model used to generate the embedding.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding_model: Option<String>,
    /// Dimensionality of the embedding vector.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub embedding_dimension: Option<u32>,

    // ── Context ──────────────────────────────────────────────────────────
    /// Programming or natural language of the content.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    /// File path, URL, or location within the source.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_path: Option<String>,
    /// Logical grouping / tenant ID for graph partitioning.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub group_id: Option<String>,

    // ── Timestamps ───────────────────────────────────────────────────────
    /// Start time for temporal content (video/audio segments).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_timestamp: Option<DateTime<Utc>>,
    /// End time for temporal content (video/audio segments).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_timestamp: Option<DateTime<Utc>>,

    // ── Rich metadata ────────────────────────────────────────────────────
    /// Content-type-specific metadata.
    #[serde(default)]
    pub metadata: EpisodeMetadata,

    // ── Provenance ───────────────────────────────────────────────────────
    /// Origin and version tracking.
    pub provenance: EpisodeProvenance,

    // ── Security / Governance ────────────────────────────────────────────
    /// Whether PII was detected in this episode's content.
    #[serde(default)]
    pub pii_detected: bool,
    /// Access control tags for graph query authorization.
    #[serde(default)]
    pub acl_tags: Vec<String>,
}

impl GraphifyEpisode {
    /// Create a new episode with sensible defaults.
    pub fn new(
        source_type: EpisodeSourceType,
        source_id: String,
        chunk_text: String,
        origin_service: &str,
    ) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            source_type,
            source_id,
            chunk_text,
            chunk_type: EpisodeChunkType::default(),
            embedding: None,
            embedding_model: None,
            embedding_dimension: None,
            language: None,
            content_path: None,
            group_id: None,
            start_timestamp: None,
            end_timestamp: None,
            metadata: EpisodeMetadata::default(),
            provenance: EpisodeProvenance {
                origin_service: origin_service.to_string(),
                version: default_version(),
                pipeline_run_id: None,
                created_at: Utc::now(),
                parent_episode_id: None,
            },
            pii_detected: false,
            acl_tags: vec![],
        }
    }

    /// Builder method: set chunk type.
    pub fn with_chunk_type(mut self, chunk_type: EpisodeChunkType) -> Self {
        self.chunk_type = chunk_type;
        self
    }

    /// Builder method: set language.
    pub fn with_language(mut self, language: impl Into<String>) -> Self {
        self.language = Some(language.into());
        self
    }

    /// Builder method: set content path.
    pub fn with_content_path(mut self, path: impl Into<String>) -> Self {
        self.content_path = Some(path.into());
        self
    }

    /// Builder method: set group ID.
    pub fn with_group_id(mut self, group_id: impl Into<String>) -> Self {
        self.group_id = Some(group_id.into());
        self
    }

    /// Builder method: attach embedding data.
    pub fn with_embedding(mut self, embedding: Vec<f32>, model: &str) -> Self {
        self.embedding_dimension = Some(embedding.len() as u32);
        self.embedding = Some(embedding);
        self.embedding_model = Some(model.to_string());
        self
    }

    /// Serialize to a Kafka-friendly JSON string.
    pub fn to_kafka_payload(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    /// Deserialize from a Kafka payload JSON string.
    pub fn from_kafka_payload(payload: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(payload)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_episode_serialization_roundtrip() {
        let episode = GraphifyEpisode::new(
            EpisodeSourceType::Codebase,
            "repo-123".to_string(),
            "fn main() {}".to_string(),
            "unified-processor",
        )
        .with_chunk_type(EpisodeChunkType::Function)
        .with_language("rust")
        .with_content_path("src/main.rs")
        .with_group_id("tenant-1");

        let json = episode.to_kafka_payload().unwrap();
        let deserialized = GraphifyEpisode::from_kafka_payload(&json).unwrap();

        assert_eq!(deserialized.id, episode.id);
        assert_eq!(deserialized.source_type, EpisodeSourceType::Codebase);
        assert_eq!(deserialized.chunk_type, EpisodeChunkType::Function);
        assert_eq!(deserialized.language, Some("rust".to_string()));
    }

    #[test]
    fn test_episode_with_embedding() {
        let episode = GraphifyEpisode::new(
            EpisodeSourceType::Document,
            "doc-456".to_string(),
            "Some document text".to_string(),
            "embeddings-service",
        )
        .with_embedding(vec![0.1, 0.2, 0.3], "all-MiniLM-L6-v2");

        assert_eq!(episode.embedding_dimension, Some(3));
        assert_eq!(episode.embedding_model.unwrap(), "all-MiniLM-L6-v2");
    }
}
